client/lib/user-settings/index.js

```diff
@@ -29,6 +26,26 @@ function decodeUserSettingsEntities( data ) {
 	return assign( {}, data, decodedValues );
 }
 
+/*
+ * Deletes a provided unsaved setting, then calls itself recursively
+ * to delete any empty parents of the setting passed to it
+ */
+function deleteUnsavedSetting( settings, settingName, recursive ) {
+	if ( ! isEmpty( get( settings, settingName ) ) || recursive ) {
+		unset( settings, settingName );
+
+		const settingKeys = settingName.split( '.' );
+		if ( settingKeys.length > 1 ) {
+			settingKeys.pop();
+			const parentKey = settingKeys.join( '.' );
+			//if parent is empty, call function again
+			if ( parentKey && isEmpty( get( settings, parentKey ) ) ) {
+				deleteUnsavedSetting( settings, parentKey, true );
```

Comment:
Yes, the idea behind the update to user settings is to support nested settings like `'preferences.colorTheme'`. The current way of deletion only deleted the key but resulted in an empty parent key in `this.unsavedSettings`. As `hasUnsavedUserSettings` only checks for `isEmpty`, it would return true despite no settings being unsaved. Therefore we have to ensure to remove potentially empty parent keys.

e.g. `{calypso_preferences: { colorScheme: "dark" }` 
=> `{calypso_preferences: { } }`
=> `{ }`